### ISSUE TITLE
Spline applicative semantics?

### ISSUE THREAD

**Issue by mitchellwrosen** [schell/varying#3]
Hi, I'm poking through the code and I have a question about the applicative semantics of SplineT. The documentation says

> It responds to <*> by applying the left arguments eventual value (the function) to the right arguments eventual value.

However, doesn't this rely on the left and right splines "ending" at the same exact time? If they don't, either the `a -> b` or the `a` will be lost to the applicative instance of `Event`, and the spline will continue.

It seems like the intended semantics of a spline are "step this (self-mutating) `a -> (Maybe b, Maybe c)` function until it emits a `Just c`, and then go from there". But, of course, this inner `Var` will continue to emit `(Maybe b, Maybe c)` pairs as long as you feed it `a`s.

So, I suppose my question is, what is the intuition behind splines? And, bonus question, have you checked the `Applicative` and `Monad` laws for it?

**Comment by schell** [comment]
The intuition behind `SplineT`'s Applicative instance _should_ be the same as other monads, like `IO`. If `ourSpline` is a spline that eventually results in an `Int` then `(+1) <$> ourSpline` will increment the result value. If ourSpline never concludes then just like in the case of an `IO` action that never concludes, the result will never be incremented.

In the case of `<*>` both splines _should_ be able to conclude at any time - they don't have to conclude at exactly the same time. My thinking is that like a `VarT`, `SplineT` is locally stateful, so it holds on to the left and right argument until both have concluded. You can see this in the Applicative instance for `SplineT`.
<img width="646" alt="screenshot 2016-03-03 09 14 00" src="https://cloud.githubusercontent.com/assets/24942/13502561/50db2be2-e120-11e5-9f1f-d88033d0c306.png">
`SplineTConst` is a spline that has concluded. You can see in the first `<*>` we only apply the function once both have concluded - otherwise either of the two running splines will be held onto.

I believe this intuition holds and I'm running through the code as I write this to make sure that at least I'm not obviously mistaken. So let's inspect the last implementation of `<*>` pasted above, where neither spline has concluded. In this case we construct a new `SplineT` that marries a modified function stream with our argument stream. The function stream is modified by having `<*>` `fmap`'d into it. `VarT`'s `fmap` implementation is as follows
<img width="556" alt="screenshot 2016-03-03 09 26 58" src="https://cloud.githubusercontent.com/assets/24942/13502990/193c448a-e122-11e5-929a-961e31724f15.png">
which shows that `fmap` creates a new `VarT` from the function and then plugs it into the given stream to transform the streams output. 

The type of a `SplineT` is really just a `VarT` that produces a special step
<img width="698" alt="screenshot 2016-03-03 09 35 39" src="https://cloud.githubusercontent.com/assets/24942/13503258/4e881fb4-e123-11e5-93f5-faf3c3637284.png">
and when the right value of the step procs (is `Event x`) the spline concludes.

At the very bottom of our interaction (after all the levels of `fmap` and `<*>`, and after unwrapping the `Step`) we have two eventual results - the function spline's eventual result and the argument spline's eventual result. Both will be produced each iteration and applied to either inhibit or produce.

The key here is that `fmap`ing a function into a `VarT` creates a new `VarT` and appends that to the original. This way neither the function nor the argument are lost - they are each bundled up in local state.

# disclamers

That said - I'm not :100:  on all that - I haven't written tests for it yet (which I'll do right now). I also didn't give a ton of thought to the monad or applicative laws - I assumed that they hold but have not written tests for that either (which I'm also doing right now).

**Comment by schell** [comment]
I realize I didn't directly inspect the other implementations of `<*>`, where one or the other spline has already concluded. Those use the same method of `fmap`ing another function onto the eventual result, which is really creating a new `VarT` and appending it to the running spline under the hood.

**Comment by mitchellwrosen** [comment]
Wow, thanks for the detailed reply, with pictures!

However, I'm not sure a spline that finishes ends up "holding on to" its final value until the other spline finishes. Take a look at the last line in the applicative instance:

``` haskell
SplineT vf <*> SplineT vx = SplineT (liftA2 (<*>) vf vx)
```

So, we first map `(<*>)` over the `vf :: VarT m a (Step (Event b) (x -> y))` to get something with type

``` haskell
VarT m a (Step (Event b) x -> Step (Event b) y)
```

This has the semantics that the `Event b` are just combined monoidally, and we only get back a `y` if both the original `x -> y` and the new `x` are `Just`.

And then we use `VarT`'s applicative instance to combine it with the

``` haskell
vx :: VarT m a (Step (Event b) x)
```

leaving us

``` haskell
VarT m a (Step (Event b) y)
```

Now, here our resulting `VarT` just steps each of the original ones and does function application.

Consider the case when stepping the first original `VarT` does indeed yield a function. Still, it never "terminates" the whole spline to a `SplineTConst` constructor, because how could it?

Maybe I'm not seeing something here. I haven't actually played around with the applicative instance yet; I've just read the code. Thanks for the reply! :)

**Comment by schell** [comment]
First off - thanks for the `liftA2` transformation - it's much nicer.

I see the confusion. I'm sorry - I explained the `SplineTConst` incorrectly. My mistake. A `SplineTConst` does not represent a spline that has concluded - it represents a spline that concluded immediately. This means it was created with `pure` or `return` - it's a constant value that never changes and has no output value since it terminates immediately. 

I'm not completely sure whether concluded splines get pruned from the network. My hope and intuition is that they do, but I should add a test for that. If you look at `SplineT`'s monad instance you'll see where that would happen.
<img width="638" alt="screenshot 2016-03-03 11 08 04" src="https://cloud.githubusercontent.com/assets/24942/13505906/39c984e8-e130-11e5-9bed-cdd9917ab8a9.png">
In the case that the previous spline produces a result value we immediately run the next spline, which will produce a discrete step and the var for the next network step. It's here that the previous spline is culled (I hope). I think I can write a test using a `SplineT a b (Writer [String]) ()` which will show when certain splines get culled or not.

**Comment by schell** [comment]
Quick update - the applicative laws hold, the monad laws do not! I'll have to figure out a way to fix this! I hope its possible.

**Comment by mitchellwrosen** [comment]
Interesting! Would you mind putting the proofs here? I'd like to see how other people work through them. I always seem to get stuck at one point or another, heh.

And re: your last comment, I believe one source of confusion is the comment about the applicative instance semantics.

> A spline is an applicative if its output type is a monoid. It responds to pure by returning a spline that immediately returns the argument. It responds to <*> by applying the left arguments eventual value (the function) to the right arguments eventual value. The output values will me combined with mappend.

I think this is not true, for reasons above, although maybe we still disagree here. It's my understanding that the expression `sf <*> sx` creates a spline `sy` such that the inner var's `b`s are the pointwise monoidal sum of `sf`'s and `sx`'s, and the inner var's `c`s are emitted when both `sf`'s and `sx`'s emitted at the same step, and is `NoEvent` otherwise.

Does that sound right? That it's basically a pointwise `(<*>)` all the way down?

**Comment by schell** [comment]
Oh wow - my documentation is outdated.There's nothing worse than no documentation... ...except for incorrect documentation. The current Applicative and Monad instances do not require that the output type be a Monoid.

I've pushed my tests to github - you should be able to run them with `stack test`. The source is [here](https://github.com/schell/varying/blob/master/test/Main.hs#L108).

**Comment by schell** [comment]
My problem now is finding another way to write the Monad instance to adhere to the Monad laws and have it type check. The type of the result value changes each sequence which makes things tricky.

**Comment by mitchellwrosen** [comment]
Maybe try a type-aligned sequence based approach?

**Comment by schell** [comment]
I'm not familiar with type alignment yet - can you link me to some info? Is it a bit like using type operators to encode a sequence of result values?

**Comment by mitchellwrosen** [comment]
Yeah, check out Oleg's paper called Freer Monads, More Extensible Effects.
It's basically an arrow with a known input type, known output type, and
existentially quantified innards that compose in the obvious way.

**Comment by schell** [comment]
Oh cool - extensible effects. I have used the extensible-effects library and I like it. I'll see what I can glean from those series of papers, thanks.

**Comment by mitchellwrosen** [comment]
Yup. I think `extensible-effects` is actually based on an older paper, but `freer` is closer to the lastest & greatest. In fact, they have a type-aligned thingy exported, too:

https://hackage.haskell.org/package/freer-0.2.2.4/docs/Data-FTCQueue.html

Not sure exactly what the problem with the monad instance is, but I'll look at the tests you mentioned and see if I can make sense of it.

**Comment by schell** [comment]
I've read the paper, played around with FTCQueue and tried a couple different implementations but I haven't had any luck with anything. 

My goal with a spline is to be able to maintain an output value that changes over time until a certain event happens and then generate the next spline using the last output value and the event value. The event value represents the result of the spline, but I'd like to be able to use both to determine the next spline.

I've tried a couple different strategies - the first is building sequences of splines from an AST using `free`. This is the easiest approach but it only lets you use the last event value and has no return value. This ends up being pretty useless for UI, which is my main usecase. 

`FTCQueue` seemed like it would work but I ran into some hurdles - the biggest being that a result value `Event c` isn't produced until a `VarT m a (Event c)` is run at least once, which is too late for generating a new spline when writing `>>=`.

The general method I used to chain splines together involved lining up a sequence of `VarT m a (Event c)` - switching whenever the current signal inhibits (produces `NoEvent`). I'm going to try an implementation from the other direction by lining up a series of `VarT m a (b, Event c)` - which is more like the current implementation. The problem with this method is that the spline result `Event c` is part of the signal - which is likely to cause the resulting Monad instance to disobey the right and left unit laws.

**Comment by mitchellwrosen** [comment]
Ok, I think I followed most of that =)

Is it possible that `Var` is the wrong abstraction for how you want it to interact with subsequent `Var`s?

For example, maybe instead of

``` haskell
type Var a b = a -> (b, Var a b)
```

it may make sense to put the concept of a "return value" into `Var` itself, as

``` haskell
type Var r a b = a -> (b, Either r (Var r a b))
```

which is starting to look similar to a free monad.

**Comment by mitchellwrosen** [comment]
Hm, I realize now this is just `VarT` with `Either r` as the monad.

**Comment by schell** [comment]
Thanks, I figured it out last night. Instead of a spline written as 

``` haskell
data SplineT a b m c = SplineT { unSplineT :: VarT m a (Step (Event b) c) }
                     | SplineTConst c
```

The trick was to write it as 

``` haskell
data SplineT a b m c = SplineT { unSplineT :: VarT m a (b, Event c) }
                     | Pass c
```

The main difference being that a step is now `(b, Event c)` instead of `(Event b, Event c)` and switching occurs when the second value in the tuple is an `Event`, instead of when the first value is `NoEvent`. This change means we lose MonadTrans and MonadIO instances, the `pair` function and the type of the remaining functions change a bit, but I think this is easier to reason about.

So long story short the key was to run a spline until a separate "switch" event procs, instead of running a spline until the output value inhibits.

**Comment by mitchellwrosen** [comment]
Cool. This seems very similar to `auto`'s [switchFromF](https://hackage.haskell.org/package/auto-0.4.3.0/docs/Control-Auto-Switch.html), except instead of 

``` haskell
switch :: (c -> Var a (b, Event c)) -> Var a (b, Event c) -> Var a b    
```

you're using the last emitted `b` to inform the switch, like

``` haskell
switch :: ((b,c) -> Var a (b, Event c)) -> Var a (b, Event c) -> Var a b    
```

**Comment by schell** [comment]
Yeah! Plus you can use do-notation :)

**Comment by schell** [comment]
I'll have the changes in tonight.

**Comment by mitchellwrosen** [comment]
I'm not _totally_ convinced the monad laws of the new `SplineT` hold, but I tried proving them by hand and got a bit stuck. I wish there was some automated monad-law prover =)

**Comment by schell** [comment]
Splines are only useful for constructing continuous streams from discontinuous pieces. That said you can turn a spline into a var using `outputStream` or `resultStream` or `runSplineT`. Using `runSplineT` will create a stream of output values AND results and so I could see `runSplineT` being a case where it doesn't abide by the monad laws, since the result values would encode each step in a do-notation, where `m >>= return` is two steps and `m` is just one. It may not though. I'll write a test for that one too.

Are there any situations you can think of that I could write tests for?

**Comment by schell** [comment]
Oh man - I guess I already wrote a test for the `runSplineT` case. In https://github.com/schell/varying/blob/master/test/Main.hs#L152

``` haskell
it "(right unit w/ monadic results) m >>= return = m" $
  (scanVar (runSplineT m 0) [0..9])
    `shouldBe` scanVar (runSplineT mr 0) [0..9]
```

and

``` haskell
let m = (1 ~> accumulate (+) 0) `untilEvent_` (1 ~> after 3)
    mr = m >>= return
```

**Comment by schell** [comment]
I updated the tests to use `untilEvent` instead of `untilEvent_`, that way the result values are kept in the var stream. The tests still pass - though I know that's not proof.

**Comment by schell** [comment]
I've added some more tests for this as well.

**Comment by mitchellwrosen** [comment]
I guess one concern I have is it doesn't seem

``` haskell
sf <*> sa
```

is equivalent to

``` haskell
f <- sf
a <- sa
return (f a)
```

because the former will step the splines in parallel with the same input, and the second will only move to the `sa` spline after `sf` emits an event.

**Comment by schell** [comment]
Huh - yeah, you're right. That's kind of by design though - to me it seemed that monads are sequential and applicatives are parallel but when you apply a monad it should work like a monad. I'll do some research. Thanks for the feedback.

**Comment by mitchellwrosen** [comment]
No problem. Maybe look at `haxl` for some inspiration. Just to be clear, though, the above behavior _does_ indicate a violation of the monad laws, I just haven't pinpointed exactly which one (probably composition). Put another way, your code can't change behavior with `{-# ApplicativeDo #-}` turned on :)

**Comment by schell** [comment]
Just for the sake of my education - is that written down anywhere? Like in Typeclassopedia or the like? 

Doesn't `{-# ApplicativeDo #-}` just sugar Applicative's syntax? ie It wouldn't actually use the monadic `>>=` and `return` but instead it would use Applicative's `(<$>, <*>)` and `pure`. Correct?

**Comment by mitchellwrosen** [comment]
Yeah, the `Control.Applicative` module says that `(<*>) = ap` is a law.

And yes - exactly! If you have `sf >>= \f -> sa >>= \a -> return (f a)` _without_ `ApplicativeDo`, it will use `>>=`, but _with_ `ApplicativeDo`, it will become `sf <*> sa`. Therefore, they have to agree on semantics, at least "observationally". (For example, `haxl`'s applicative semantics delay the actual IO fetching of  requested data until a bona fide `>>=` is required).

**Comment by schell** [comment]
Okay - one last thing - if those two statements should be equivalent by law, and all Monads must also be Applicatives - couldn't all Monad's Applicative instances have 

``` haskell
do f <- mf
   x <- mx
   return $ f x
```

as the default implementation for `<*>` ?

**Comment by schell** [comment]
It seems to work, at least. And the tests hold.

**Comment by schell** [comment]
@mitchellwrosen I didn't expect that to auto-close but I'll reopen if you find anything else.

**Comment by mitchellwrosen** [comment]
Yeah, that definition for `<*>` is correct assuming the monad laws hold. I've been playing around with [prover](https://github.com/feuerbach/prover) a bit, just for fun:

``` haskell
-- type Var a b = \/r. ((a -> (\/s. (b -> r -> s) -> s)) -> r) -> r
--
-- or
--
-- type Pair a b = \/r. (a -> b -> r) -> r
-- type Var a b = \/r. ((a -> Pair b r) -> r) -> r

-- map :: (b -> c) -> Var a b -> Var a c
map = \f. \v. \k. v (\g. k (\a. \c. g a (\b. \x. c (f b) x)));
```

And with it, I've proven `fmap` for `VarT Identity`:

```
> prover -i var.lam --equal 'map (\a. a) x' 'x' --fuel 50
> prover -i var.lam --equal 'map f (map g x)' 'map (\a. f (g a)) x' --fuel 50
```

Obviously, the functor laws are so simple that the GHC can derive them, but I hope proving applicative and monad laws will be possible. I think I have `pure`, too:

``` haskell
fix = \f. f (fix f);
pure = \b. \k. fix (\r. k (\a. \g. g b r));
```

**Comment by schell** [comment]
Wow! Cool! I'll have to check out `prover` when I get a second.

**Comment by schell** [comment]
Using the Monad instance's sequencing for Applicative has the added benefit that we don't lose all the output values of the left stream as well, so thanks - I didn't like that part of my implementation.

**Comment by mitchellwrosen** [comment]
I got super stuck trying to write `(<*>)` in the untyped lambda calc, heh. Church encoding is confusing - I'm not even totally sure this is correct:

``` haskell
type Var a b = forall r. ((a -> (forall s. (b -> r -> s) -> s)) -> r) -> r
```

(which is what I proved `fmap` for)

**Comment by schell** [comment]
I wouldn't know either! Thanks for your help nonetheless.
